rhp/v4/rpc_test.go

```diff
@@ -378,70 +377,119 @@ func TestRenewContractPartialRollover(t *testing.T) {
+			Allowance:       renterAllowance,
+			Collateral:      hostCollateral,
+			ProofHeight:     cm.Tip().Height + 50,
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+		revision := result.Contract
 
-	// verify the transaction set is valid
-	if known, err := cm.AddV2PoolTransactions(result.FormationSet.Basis, result.FormationSet.Transactions); err != nil {
-		t.Fatal(err)
-	} else if !known {
-		t.Fatal("expected transaction set to be known")
-	}
+		// verify the transaction set is valid
+		if known, err := cm.AddV2PoolTransactions(result.FormationSet.Basis, result.FormationSet.Transactions); err != nil {
+			t.Fatal(err)
+		} else if !known {
+			t.Fatal("expected transaction set to be known")
+		}
 
-	// mine a few blocks to confirm the contract
-	mineAndSync(t, cm, types.VoidAddress, 10, w, c)
+		// mine a few blocks to confirm the contract
+		mineAndSync(t, cm, types.VoidAddress, 10, w, c)
 
-	// fund an account to transfer funds to the host
-	cs := cm.TipState()
-	account := proto4.Account(renterKey.PublicKey())
-	accountFundAmount := types.Siacoins(25)
-	fundResult, err := rhp4.RPCFundAccounts(context.Background(), transport, cs, renterKey, revision, []proto4.AccountDeposit{
-		{Account: account, Amount: accountFundAmount},
+		// fund an account to transfer funds to the host
+		cs := cm.TipState()
+		account := proto4.Account(renterKey.PublicKey())
+		accountFundAmount := types.Siacoins(25)
+		fundResult, err := rhp4.RPCFundAccounts(context.Background(), transport, cs, renterKey, revision, []proto4.AccountDeposit{
+			{Account: account, Amount: accountFundAmount},
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+		revision.Revision = fundResult.Revision
+
+		// renew the contract
+		_, err = rhp4.RPCRefreshContract(context.Background(), transport, cm, fundAndSign, cm.TipState(), settings.Prices, revision.Revision, proto4.RPCRefreshContractParams{
+			ContractID: revision.ID,
+			Allowance:  types.ZeroCurrency,
+			Collateral: types.ZeroCurrency,
+		})
+		if err == nil {
+			t.Fatal(err)
+		} else if !strings.Contains(err.Error(), "allowance is zero") {
+			t.Fatalf("unexpected error: %v", err)
+		}
 	})
-	if err != nil {
-		t.Fatal(err)
-	}
-	revision.Revision = fundResult.Revision
 
-	// renew the contract
-	renewResult, err := rhp4.RPCRenewContract(context.Background(), transport, cm, fundAndSign, cm.TipState(), settings.Prices, revision.Revision, proto4.RPCRenewContractParams{
-		ContractID:  revision.ID,
-		Allowance:   types.Siacoins(150),
-		Collateral:  types.Siacoins(300),
-		ProofHeight: revision.Revision.ProofHeight + 10,
-	})
-	if err != nil {
-		t.Fatal(err)
-	}
+	t.Run("valid refresh", func(t *testing.T) {
+		fundAndSign := &fundAndSign{w, renterKey}
+		renterAllowance, hostCollateral := types.Siacoins(100), types.Siacoins(200)
+		result, err := rhp4.RPCFormContract(context.Background(), transport, cm, fundAndSign, cm.TipState(), settings.Prices, hostKey.PublicKey(), settings.WalletAddress, proto4.RPCFormContractParams{
```

Comment:
extracting these ~40 lines into a test helper that forms a contract and verifies the transaction set is valid, perhaps also funds an account would save a lot of lines in this file, seems like these steps are repeated 5-6 times over and are exact copies